File: dataModule/src/main/java/com/mobnetic/coinguardian/model/market/ItBit.java
package com.mobnetic.coinguardian.model.market

import com.mobnetic.coinguardian.model.CheckerInfo
import com.mobnetic.coinguardian.model.Market
import com.mobnetic.coinguardian.model.Ticker
import com.mobnetic.coinguardian.model.currency.Currency
import com.mobnetic.coinguardian.model.currency.CurrencyPairsMap
import com.mobnetic.coinguardian.model.currency.VirtualCurrency
import org.json.JSONObject
import java.util.*

class ItBit : Market(NAME, TTS_NAME, CURRENCY_PAIRS) {
    companion object {
        private const val NAME = "itBit"
        private const val TTS_NAME = "It Bit"
        private const val URL = "https://api.itbit.com/v1/markets/%1\$s%2\$s/ticker"
        private val CURRENCY_PAIRS: CurrencyPairsMap = CurrencyPairsMap()

        init {
            CURRENCY_PAIRS[VirtualCurrency.BTC] = arrayOf(
                    Currency.USD,
                    Currency.SGD,
                    Currency.EUR
            )
        }
    }

    private fun fixCurrency(currency: String?): String? {
        return if (VirtualCurrency.BTC == currency) VirtualCurrency.XBT else currency
    }

    override fun getUrl(requestId: Int, checkerInfo: CheckerInfo): String {
        return String.format(URL, fixCurrency(checkerInfo.currencyBase), checkerInfo.currencyCounter)
    }

    @Throws(Exception::class)
    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {
        ticker.bid = jsonObject.getDouble("bid")
        ticker.ask = jsonObject.getDouble("ask")
        ticker.vol = jsonObject.getDouble("volume24h")
        ticker.high = jsonObject.getDouble("high24h")
        ticker.low = jsonObject.getDouble("low24h")
        ticker.last = jsonObject.getDouble("lastPrice")
    }

    @Throws(Exception::class)
    override fun parseErrorFromJsonObject(requestId: Int, jsonObject: JSONObject, checkerInfo: CheckerInfo?): String? {
        return jsonObject.getString("message") // not working?
    }
}

File: dataModule/src/main/java/com/aneonex/bitcoinchecker/datamodule/model/market/ItBit.kt
package com.aneonex.bitcoinchecker.datamodule.model.market

import com.aneonex.bitcoinchecker.datamodule.model.CheckerInfo
import com.aneonex.bitcoinchecker.datamodule.model.CurrencyPairInfo
import com.aneonex.bitcoinchecker.datamodule.model.Ticker
import com.aneonex.bitcoinchecker.datamodule.model.currency.VirtualCurrency
import com.aneonex.bitcoinchecker.datamodule.model.market.generic.SimpleMarket
import com.aneonex.bitcoinchecker.datamodule.util.forEachJSONObject
import org.json.JSONObject

class ItBit : SimpleMarket(
    "itBit (by Paxos)",
    "https://api.paxos.com/v2/markets",
    "https://api.paxos.com/v2/markets/%1\$s/ticker",
    "It Bit"
) {
    private fun fixCurrency(currency: String?): String? {
        return if (VirtualCurrency.BTC == currency) VirtualCurrency.XBT else currency
    }

    override fun getPairId(checkerInfo: CheckerInfo): String {
        return checkerInfo.currencyPairId ?: "${fixCurrency(checkerInfo.currencyBase)}{checkerInfo.currencyCounter}"
    }

    override fun parseCurrencyPairsFromJsonObject(
        requestId: Int,
        jsonObject: JSONObject,
        pairs: MutableList<CurrencyPairInfo>
    ) {
        jsonObject
            .getJSONArray("markets")
            .forEachJSONObject {
                pairs.add(CurrencyPairInfo(
                    it.getString("base_asset"),
                    it.getString("quote_asset"),
                    it.getString("market")
                ))
            }
    }

    override fun parseTickerFromJsonObject(requestId: Int, jsonObject: JSONObject, ticker: Ticker, checkerInfo: CheckerInfo) {

        ticker.bid = jsonObject.getJSONObject("best_bid").getDouble("price")
        ticker.ask = jsonObject.getJSONObject("best_ask").getDouble("price")

        jsonObject.getJSONObject("last_day").also {
            ticker.vol = it.getDouble("volume")
            ticker.high = it.getDouble("high")
            ticker.low = it.getDouble("low")
        }

        ticker.last = jsonObject.getJSONObject("last_execution").getDouble("price")
    }

    override fun parseErrorFromJsonObject(
        requestId: Int,
        jsonObject: JSONObject,
        checkerInfo: CheckerInfo
    ): String? {
        return jsonObject
            .getJSONObject("error")
            .getString("message")
    }
}